etcd operator working group (#7917)
kubernetes/community: wg-etcd-operator/charter.md

# WG etcd operator

This charter adheres to the conventions described in the [Kubernetes Charter README]
and uses the Roles and Organization Management outlined in [sig-governance].

[sig-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-governance.md
[Kubernetes Charter README]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/README.md

## Scope

The purpose of an etcd-operator is to operate automatically etcd clusters which run in the Kubernetes environment.
It minimizes human intervention as much as possible.

### In Scope

- Collect requirements & use cases with a [survey](https://forms.gle/5gBpzaxYtuQPWdBo9) to better understand what users care about the most.
- Prioritize the tasks based on feedback and create a roadmap.
- Bootstrap a project "etcd-operator" owned by SIG etcd which resides in the etcd-io or kubernetes-sigs Github orgs.
  - Review existing etcd operators to see if any could be forked or referenced to advance the project.

> - Discuss & design the core reconciliation workflow, and potentially provide a proof of concept (PoC).

Author: updated & thx

- Figure out how to get resource for following dev/test, i.e. AWS S3.

### Out of scope

- Manage etcd clusters running within non-Kubernetes environments.

> - Manage etcd clusters which are supporting the Kubernetes itself.

Reviewer: ```suggestion
- Manage etcd clusters which are used as the storage backend of a host (non-nested) kube-apiservers.
```

trying to explain this better.

Author: updated & thx

## Stakeholders

Stakeholders for this working group include members in the following SIGs:

- SIG etcd
- SIG Cluster Lifecycle

## Deliverables

The artifacts the group is supposed to deliver include:

> - Survey results which describe the users requirements and use cases.

Author: Updated & thx

- Roadmap for the project etcd-operator.
- Core reconciliation workflow and PoC.

> - A new repository "etcd-operator" owned by SIG etcd, and it should have implemented the very basic functionalities below,
>   - Creation of a new etcd cluster with one or multiple members;
>   - Scale out & in the etcd cluster;
>   - Upgrading patch versions or one minor version.

Reviewer: ```suggestion
- A new repository "etcd-operator" owned by SIG etcd, and it should have implemented the very basic functionalities below:
  - Creation of a new etcd cluster with one or multiple members.
  - Scale in and out the etcd cluster.
  - Upgrading patch and minor versions.
```

Author: Partially updated, thx. 

`Upgrading patch versions or one minor version` seems better, because we support upgrade across multiple patch version, but only one minor version.

# WG etcd operator

This charter adheres to the conventions described in the [Kubernetes Charter README]
and uses the Roles and Organization Management outlined in [sig-governance].

[sig-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-governance.md
[Kubernetes Charter README]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/README.md

## Scope

> The purpose of an etcd-operator is to operate automatically etcd clusters which run in the Kubernetes environment.

Reviewer 2: IIUC one of the main goals is to cover only operating etcd, but not the one backing the k8s cluster. I think it's important to make this explicit, so that users won't get confused about it. I'm aware it's mentioned as out of scope, but let's repeat it here. I remember when I first heard about it, that particular etcd was what I thought about.

> The purpose of an etcd-operator is to operate automatically etcd clusters which run in the Kubernetes environment.

Reviewer 2: Maybe something in the lines of:
```suggestion
The purpose of an etcd-operator is to operate automatically etcd clusters which run in the Kubernetes environment (with the exclusion of the etcd backing kubernetes cluster).
```

Author: Thanks for the suggestion. Updated, it should be clearer now. PTAL.

It minimizes human intervention as much as possible.

### In Scope

- Collect requirements & use cases with a [survey](https://forms.gle/5gBpzaxYtuQPWdBo9) to better understand what users care about the most.
- Prioritize the tasks based on feedback and create a roadmap.
- Bootstrap a project "etcd-operator" owned by SIG etcd which resides in the etcd-io or kubernetes-sigs Github orgs.
  - Review existing etcd operators to see if any could be forked or referenced to advance the project.
- Discuss and design the core reconciliation workflow, and potentially provide a proof of concept (PoC).

> - Figure out how to get resource for following dev/test, i.e. AWS S3.

Reviewer 3: Not sure if this was already discussed, would a possible integration with Cluster API in SIG Cluster Lifecycle also be in scope? Can be as much as "it works with as an addon" or provider of sorts. It'd be great to just keep it in mind for native support of external etcd w/ kubeadm

cc @fabriziopandini @sbueringer

Author: > It'd be great to just keep it in mind for native support of external etcd w/ kubeadm

thx for the comment. It's already been discussed. It might be a potential long-term goal, but not in the scope for now. The intention is to keep it as simple as possible in the first place. It's also clarified in the section "Out of scope" (see below),

```
### Out of scope

- Manage etcd clusters running within non-Kubernetes environments.
- Manage etcd clusters which are used as the storage backend of a host (non-nested) kube-apiserver.
```

Reviewer 4: I think CAPI would be doing neither of those. CAPI would be using it roughly like any other application.

We would host etcd on one Kubernetes cluster and the apiserver of another cluster would use it as backend. So no chicken-egg problems I think

Author: thx for the info.

Seems like the use case is similar to the [Hosted Control Plane](https://kccnceu2024.sched.com/event/1YeQ1).

The etcd cluster, which is managed by the etcd-operator, will run as PODs within Kubernetes environment, treating them like any other typical applications. I think we are open to any use cases on top of this base.

Regarding integrating with Cluster API to support the use case of external etcd with Kubeadm, we need cluster API maintainers' help to make it work, and it's open to get it included in the roadmap (pls see the section "In Scope") of the subproject.

Reviewer 5: I think, it's responsibility of the ControlPlane Provider to deploy etcd, not the CAPI manager itself. That's how we do in in [k0smotron](https://github.com/k0sproject/k0smotron/) ControlPlane Provider, other implementations do more or less the same.

- Manage etcd clusters running within non-Kubernetes environments.
- Manage etcd clusters which are used as the storage backend of a host (non-nested) kube-apiserver.

## Stakeholders

Stakeholders for this working group include members in the following SIGs:

- SIG etcd
- SIG Cluster Lifecycle

## Deliverables

The artifacts the group is supposed to deliver include:
- Survey results which describe the users requirements and use cases.
- Evaluation results on existing etcd-operators.
- Roadmap for the project etcd-operator.
- Core reconciliation workflow and PoC.
- A new repository "etcd-operator" owned by SIG etcd, and it should have implemented the very basic functionalities below:
  - Creation of a new etcd cluster with one or multiple members.
  - Scale out and in the etcd cluster.
  - Upgrading patch versions or one minor version.

## Roles and Organization Management

This working group adheres to the Roles and Organization Management outlined in
[sig-governance] and opts-in to updates and modifications to [sig-governance].

[sig-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-governance.md

## Timelines and Disbanding

> The exact timeline for this working group is hard to define at this time.

Reviewer 2: A lot of our WGs have a hard time figuring out the exact timing. The next paragraph is sufficient to match the requirements, though. So I'd suggest dropping this statement and just go with the next one, only.

Author: Removed

When all the deliverables mentioned above are done and there is no additional coordination needed,

> then we will disband this working group and continue to track the development of the proejct

Reviewer 6: nit: s/proejct/project

Author: updated & thx
